Build failure with GHC 9.10

Some change to the GHC internal syntax...
```
src/Test/DocTest/Internal/Extract.hs:333:3: error: [GHC-27346]
    • The data constructor ‘HsPar’ should have 2 arguments, but has been given 4
    • In the pattern: HsPar _ _ (L l e) _
      In a \case alternative:
          HsPar _ _ (L l e) _ -> extractLit (locA l) e
      In the expression:
        \case
          HsPar _ _ (L l e) _ -> extractLit (locA l) e
          ExprWithTySig _ (L l e) _ -> extractLit (locA l) e
          HsOverLit _ OverLit {ol_val = HsIsString _ s}
            -> Just (toLocated (L loc (unpackFS s)))
          HsLit _ (HsString _ s) -> Just (toLocated (L loc (unpackFS s)))
          _ -> Nothing
    |
333 |   HsPar _ _ (L l e) _ -> extractLit (locA l) e
    |   ^^^^^^^^^^^^^^^^^^^
```
https://github.com/haskell-hvr/regex-tdfa/actions/runs/8306825242/job/23690105630?pr=59#step:17:124

Hmm, I can't actually get GHC 9.10 running on my system. I usually install it using GHCUp:

```
ghcup install ghc -u https://downloads.haskell.org/ghc/9.10.1-alpha2/ghc-9.10.0.20240328-x86_64-deb11-linux.tar.xz ghc-9.10.1-alpha2
```

but I'm getting `ghc-configure.log`:

```
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking target system type... Invalid configuration `ghc': machine `ghc-unknown' not recognized
configure: error: /bin/bash ./config.sub ghc failed
```

and `ghcup.log`:

```
Debug: Identified Platform as: Linux Ubuntu, 22.04
Debug: last access was 281.57330577s ago, cache interval is 300s
Debug: Decoding yaml at: /home/martijn/.ghcup/cache/ghcup-0.0.8.yaml
Debug: Requested to install GHC with ghc-9.10.1-alpha2
Info: downloading: https://downloads.haskell.org/ghc/9.10.1-alpha2/ghc-9.10.0.20240328-x86_64-deb11-linux.tar.xz as file /home/martijn/.ghcup/tmp/ghcup-3b1c7df75b51c314/ghc-9.10.0.20240328-x86_64-deb11-linux.tar.xz
Info: Unpacking: ghc-9.10.0.20240328-x86_64-deb11-linux.tar.xz to /home/martijn/.ghcup/tmp/ghcup-564dd8fdc579ab75
Info: Installing GHC (this may take a while)
Debug: Running sh with arguments ["./configure","--prefix=/home/martijn/.ghcup/ghc/ghc-9.10.1-alpha2","--target=ghc"]
Error: [GHCup-00841] Process "sh" with arguments ["./configure",
                             "--prefix=/home/martijn/.ghcup/ghc/ghc-9.10.1-alpha2",
                             "--target=ghc"] failed with exit code 1.
Error: Also check the logs in /home/martijn/.ghcup/logs
```

How did you install it?

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -69,11 +69,11 @@ jobs:
           - os: windows-latest
             ghc: 8.8.4
 
-        # # Uncomment if testing with an unreleased GHC. Make sure to edit the
-        # #  "Setup Haskell (head)" step too.
-        # include:
-        #   - os: ubuntu-22.04
-        #     ghc: head
+        # Uncomment if testing with an unreleased GHC. Make sure to edit the
+        #  "Setup Haskell (head)" step too.
+        include:
+          - os: ubuntu-22.04
+            ghc: head
       fail-fast: false
     steps:
       - name: Checkout
@@ -93,8 +93,9 @@ jobs:
           sudo apt-get update
           sudo apt-get install -y build-essential curl libffi-dev libffi8ubuntu1 libgmp-dev libgmp10 libncurses-dev libncurses5 libtinfo5
           curl --proto '=https' --tlsv1.2 -sSf https://get-ghcup.haskell.org | sh
-          ghcup install -u https://downloads.haskell.org/ghc/9.6.1/ghc-9.6.1-x86_64-deb9-linux.tar.xz ghc-head
-          ghcup set ghc ghc-head
+          ghcup config add-release-channel https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-prereleases-0.0.8.yaml
+          ghcup install ghc 9.10.0.20240328
+          ghcup set ghc 9.10.0.20240328
           echo "cabal-store=$HOME/.cabal/store" >> $GITHUB_OUTPUT
 
       - name: Setup CI

--- a/CHANGES.markdown
+++ b/CHANGES.markdown
@@ -1,3 +1,6 @@
+# 0.3.1.1
+ * Add support for GHC 9.10
+
 # 0.3.1
  * Add support for GHC 9.8
  * Drop support for GHC 8.2

--- a/cabal.project
+++ b/cabal.project
@@ -3,7 +3,12 @@ packages:
 
 write-ghc-environment-files: always
 
+tests: true
+
+source-repository-package
+  type: git
+  location: https://github.com/haskell-unordered-containers/unordered-containers.git
+  tag: d52a0fd10bfa701cbbc9d7ac06bd7eb7664b3972
 
 allow-newer:
-    *:base
-  , *:ghc-bignum
+  unordered-containers:template-haskell

--- a/doctest-parallel.cabal
+++ b/doctest-parallel.cabal
@@ -1,7 +1,7 @@
 cabal-version: 2.0
 
 name:           doctest-parallel
-version:        0.3.1
+version:        0.3.1.1
 synopsis:       Test interactive Haskell examples
 description:    The doctest program checks examples in source code comments.  It is modeled
                 after doctest for Python (<https://docs.python.org/3/library/doctest.html>).
@@ -26,6 +26,7 @@ tested-with:
   , GHC == 9.4.7
   , GHC == 9.6.3
   , GHC == 9.8.1
+  , GHC == 9.10.1
 
 extra-source-files:
     example/example.cabal
@@ -102,7 +103,7 @@ library
     , directory
     , exceptions
     , filepath
-    , ghc >=8.2 && <9.9
+    , ghc >=8.2 && <9.11
     , ghc-paths >=0.1.0.9
     , process
     , random >= 1.2

--- a/example/cabal.project
+++ b/example/cabal.project
@@ -7,3 +7,11 @@ write-ghc-environment-files: always
 allow-newer:
     *:base
   , *:ghc-bignum
+
+source-repository-package
+  type: git
+  location: https://github.com/haskell-unordered-containers/unordered-containers.git
+  tag: d52a0fd10bfa701cbbc9d7ac06bd7eb7664b3972
+
+allow-newer:
+  unordered-containers:template-haskell

--- a/src/Test/DocTest/Internal/Extract.hs
+++ b/src/Test/DocTest/Internal/Extract.hs
@@ -329,8 +329,10 @@ extractLit loc = \case
 #else
 #if __GLASGOW_HASKELL__ < 904
   HsPar _ (L l e) -> extractLit (locA l) e
-#else
+#elif __GLASGOW_HASKELL__ < 909
   HsPar _ _ (L l e) _ -> extractLit (locA l) e
+#else
+  HsPar _ (L l e) -> extractLit (locA l) e
 #endif
 #if __GLASGOW_HASKELL__ < 807
   ExprWithTySig _ (L l e) -> extractLit l e

--- a/src/Test/DocTest/Internal/Interpreter.hs
+++ b/src/Test/DocTest/Internal/Interpreter.hs
@@ -1,5 +1,9 @@
 {-# LANGUAGE CPP #-}
 
+#if __GLASGOW_HASKELL__ <= 906
+{-# LANGUAGE LambdaCase #-}
+#endif
+
 module Test.DocTest.Internal.Interpreter (
   Interpreter
 , safeEval
@@ -18,6 +22,11 @@ import System.Directory (getPermissions, executable)
 import Control.Monad
 import Control.Exception hiding (handle)
 import Data.Char
+#if __GLASGOW_HASKELL__ > 906
+import Data.List (unsnoc)
+#else
+import Data.Bifunctor (first)
+#endif
 import GHC.Paths (ghc)
 
 import Test.DocTest.Internal.GhciWrapper
@@ -26,6 +35,23 @@ import Test.DocTest.Internal.Logging (DebugLogger)
 -- $setup
 -- >>> import Test.DocTest.Internal.GhciWrapper (eval)
 -- >>> import Test.DocTest.Internal.Logging (noLogger)
+
+#if __GLASGOW_HASKELL__ <= 906
+-- | If the list is empty returns 'Nothing', otherwise returns the 'init' and the 'last'.
+--
+-- > unsnoc "test" == Just ("tes",'t')
+-- > unsnoc ""     == Nothing
+-- > \xs -> unsnoc xs == if null xs then Nothing else Just (init xs, last xs)
+unsnoc :: [a] -> Maybe ([a], a)
+unsnoc = \case
+  []   -> Nothing
+  x:xs -> Just $ unsnoc1 x xs
+  where
+    unsnoc1 :: a -> [a] -> ([a], a)
+    unsnoc1 x = \case
+      []   -> ([], x)
+      y:ys -> first (x:) $ unsnoc1 y ys
+#endif
 
 haveInterpreterKey :: String
 haveInterpreterKey = "Have interpreter"
@@ -76,13 +102,13 @@ safeEvalIt repl = either (return . Left) (fmap Right . evalIt repl) . filterExpr
 
 filterExpression :: String -> Either String String
 filterExpression e =
-  case lines e of
+  case map strip (lines e) of
     [] -> Right e
-    l  -> if firstLine == ":{" && lastLine /= ":}" then fail_ else Right e
-      where
-        firstLine = strip $ head l
-        lastLine  = strip $ last l
-        fail_ = Left "unterminated multiline command"
+    (firstLine:ls) ->
+      let lastLine = maybe firstLine snd (unsnoc ls) in
+      if firstLine == ":{" && lastLine /= ":}" then fail_ else Right e
   where
+    fail_ = Left "unterminated multiline command"
+
     strip :: String -> String
     strip = dropWhile isSpace . reverse . dropWhile isSpace . reverse

--- a/src/Test/DocTest/Internal/Runner/Example.hs
+++ b/src/Test/DocTest/Internal/Runner/Example.hs
@@ -27,7 +27,7 @@ mkResult expected_ actual_ =
     -- use show to escape special characters in output lines if any output line
     -- contains any unsafe character
     escapeOutput
-      | any (not . isSafe) $ concat (expectedAsString ++ actual_) = init . tail . show . stripEnd
+      | any (not . isSafe) $ concat (expectedAsString ++ actual_) = init . drop 1 . show . stripEnd
       | otherwise = id
 
     actual :: [String]

--- a/test/PropertySpec.hs
+++ b/test/PropertySpec.hs
@@ -46,7 +46,7 @@ spec = do
       runProperty repl "x == 23" `shouldReturn` Failure "*** Failed! Falsified (after 1 test):\n0"
 
     it "reports the values for which a property that takes multiple arguments fails" $ withInterpreter noLogger [] $ \repl -> do
-      let vals x = case x of (Failure r) -> tail (lines r); _ -> error "Property did not fail!"
+      let vals x = case x of (Failure r) -> drop 1 (lines r); _ -> error "Property did not fail!"
       vals `fmap` runProperty repl "x == True && y == 10 && z == \"foo\"" `shouldReturn` ["False", "0", show ("" :: String)]
 
     it "defaults ambiguous type variables to Integer" $ withInterpreter noLogger [] $ \repl -> do

